### PR TITLE
improve(gas-golf): Minor gas tweaks

### PR DESCRIPTION
**Pull request by nicholaspai**
Makes modest gas cost reductions, I think its easy to argue we don't need these optimizations but curious what you think. 

Unit test gas report from running `yarn test:gas-analytics`

Stress testing results
```
BEFORE:
  Gas Analytics: HubPool Root Bundle Execution
    Pool Rebalance tree with 10 leaves, each containing refunds for 10 different tokens
proposeRootBundle-gasUsed: 110198
      ✔ Simple proposal
executeRootBundle-gasUsed: 714268
      ✔ Executing 1 leaf (220ms)
(average) executeRootBundle-gasUsed: 686317
      ✔ Executing all leaves (1838ms)
(average) executeRootBundle-gasUsed: 497583
      ✔ Executing all leaves using multicall (1506ms)
executeRootBundle-gasUsed: 10362678
      ✔ Stress Test: 1 leaf contains 100 L1 tokens with netSendAmounts > 0 (28541ms)

  Gas Analytics: SpokePool Deposits
    ERC20 Deposits
deposit-gasUsed: 70672
      ✔ 1 Deposit (39ms)
(average) deposit-gasUsed: 70672
      ✔ 10 deposits (264ms)
(average) deposit-gasUsed: 24380
      ✔ 10 deposits using multicall (117ms)
    WETH Deposits
deposit-gasUsed: 62709
      ✔ 1 ETH Deposit

  Gas Analytics: SpokePool Relays
    ERC20 Relays
fillRelay-gasUsed: 83390
      ✔ 1 Relay
(average) fillRelay-gasUsed: 83400
      ✔ 10 Relays (200ms)
(average) fillRelay-gasUsed: 47303
      ✔ 10 relays using multicall (110ms)
    WETH Relays
fillRelay-gasUsed: 98273
      ✔ 1 Relay

  Gas Analytics: SpokePool Relayer Refund Root Execution
    (ERC20) Tree with 10 leaves, each containing 10 refunds
relayRootBundle-gasUsed: 75995
      ✔ Relay proposal
executeRelayerRefundLeaf-gasUsed: 147265
      ✔ Executing 1 leaf (73ms)
(average) executeRelayerRefundLeaf-gasUsed: 131555
      ✔ Executing all leaves (522ms)
(average) executeRelayerRefundLeaf-gasUsed: 102826
      ✔ Executing all leaves using multicall (423ms)
    (WETH): Relayer Refund tree with 10 leaves, each containing 10 refunds
executeRelayerRefundLeaf-gasUsed: 165885
      ✔ Executing 1 leaf (74ms)

  Gas Analytics: SpokePool Slow Relay Root Execution
    (ERC20) Tree with 10 leaves
relayRootBundle-gasUsed: 75995
      ✔ Relay proposal
executeSlowRelayLeaf-gasUsed: 84868
      ✔ Executing 1 leaf
(average) executeSlowRelayLeaf-gasUsed: 99937
      ✔ Executing all leaves (211ms)
(average) executeSlowRelayLeaf-gasUsed: 75462
      ✔ Executing all leaves using multicall (135ms)
    (WETH) Tree with 10 leaves
executeSlowRelayLeaf-gasUsed: 92760
      ✔ Executing 1 leaf

AFTER
  Gas Analytics: HubPool Root Bundle Execution
    Pool Rebalance tree with 10 leaves, each containing refunds for 10 different tokens
proposeRootBundle-gasUsed: 106555
      ✓ Simple proposal
executeRootBundle-gasUsed: 712012
      ✓ Executing 1 leaf
(average) executeRootBundle-gasUsed: 684081
      ✓ Executing all leaves
(average) executeRootBundle-gasUsed: 495349
      ✓ Executing all leaves using multicall
executeRootBundle-gasUsed: 10348429
      ✓ Stress Test: 1 leaf contains 100 L1 tokens with netSendAmounts > 0

  Gas Analytics: SpokePool Deposits
    ERC20 Deposits
deposit-gasUsed: 70581
      ✓ 1 Deposit
(average) deposit-gasUsed: 70581
      ✓ 10 deposits
(average) deposit-gasUsed: 24289
      ✓ 10 deposits using multicall
    WETH Deposits
deposit-gasUsed: 62618
      ✓ 1 ETH Deposit

  Gas Analytics: SpokePool Relays
    ERC20 Relays
fillRelay-gasUsed: 83390
      ✓ 1 Relay
(average) fillRelay-gasUsed: 83400
      ✓ 10 Relays
(average) fillRelay-gasUsed: 47303
      ✓ 10 relays using multicall
    WETH Relays
fillRelay-gasUsed: 98273
      ✓ 1 Relay

  Gas Analytics: SpokePool Relayer Refund Leaf Execution
    (ERC20) Tree with 10 leaves, each containing 10 refunds
relayRootBundle-gasUsed: 75992
      ✓ Relay proposal
executeRelayerRefundLeaf-gasUsed: 146346
      ✓ Executing 1 leaf
(average) executeRelayerRefundLeaf-gasUsed: 130636
      ✓ Executing all leaves
(average) executeRelayerRefundLeaf-gasUsed: 101907
      ✓ Executing all leaves using multicall
    (WETH): Relayer Refund tree with 10 leaves, each containing 10 refunds
executeRelayerRefundLeaf-gasUsed: 164966
      ✓ Executing 1 leaf

  Gas Analytics: SpokePool Slow Relay Root Execution
    (ERC20) Tree with 10 leaves
relayRootBundle-gasUsed: 75992
      ✓ Relay proposal
executeSlowRelayLeaf-gasUsed: 84868
      ✓ Executing 1 leaf
(average) executeSlowRelayLeaf-gasUsed: 99937
      ✓ Executing all leaves
(average) executeSlowRelayLeaf-gasUsed: 75462
      ✓ Executing all leaves using multicall
    (WETH) Tree with 10 leaves
executeSlowRelayLeaf-gasUsed: 92760
      ✓ Executing 1 leaf
```

Changes
- In for loops: 
    - Pre-increment saves a Store operation
    - Cache length before loop
    - Use unchecked to increment counter
- Use `calldata` instead of `memory` for variables we don't change
- Pack state variables smarter
- Don't read from state when emitting variables

### CODE REVIEW

**Review by chrismaree — APPROVED** [review]
🏌️

**Review by mrice32 — COMMENTED** [review]
As discussed IRL, I think we might want to hold off on merging this until we decide whether we think there's enough time for OZ to review this along with other fixes.

File: contracts/HubPool.sol [hunk 1]
```diff
@@ -327,7 +327,7 @@ contract HubPool is HubPoolInterface, Testable, Lockable, MultiCaller, Ownable {
         // The bond should be the passed in bondAmount + the final fee.
         bondToken = newBondToken;
         bondAmount = newBondAmount + _getBondTokenFinalFee();
-        emit BondSet(address(newBondToken), bondAmount);
```
Comment (nicholaspai) [inline contracts/HubPool.sol:330]: Don't SLOAD unnecessarily

File: contracts/HubPool.sol [hunk 2]
```diff
@@ -889,6 +889,11 @@ contract HubPool is HubPoolInterface, Testable, Lockable, MultiCaller, Ownable {
 
             // Allocate LP fees and protocol fees from the bundle to the associated pooled token trackers.
             _allocateLpAndProtocolFees(l1Token, bundleLpFees[i]);
+
+            // L1 tokens length won't be > types(uint32).length
+            unchecked {
+                ++i;
```
Comment (nicholaspai) [inline contracts/HubPool.sol:895]: pre-increment saves a STORE vs post-increment

File: contracts/Polygon_SpokePool.sol [hunk 3]
```diff
@@ -87,7 +87,7 @@ contract Polygon_SpokePool is IFxMessageProcessor, SpokePool {
      */
     function setFxChild(address newFxChild) public onlyAdmin {
         fxChild = newFxChild;
-        emit SetFxChild(fxChild);
+        emit SetFxChild(newFxChild);
```
Comment (nicholaspai) [inline contracts/Polygon_SpokePool.sol:90]: Don't SLOAD unnecessarily

File: contracts/HubPool.sol [hunk 4]
```diff
@@ -71,9 +71,6 @@ contract HubPool is HubPoolInterface, Testable, Lockable, MultiCaller, Ownable {
     // can be submitted.
     RootBundle public rootBundleProposal;
 
-    // Whether the bundle proposal process is paused.
```
Comment (nicholaspai) [inline contracts/HubPool.sol:74]: Any changes in the order of state variables is done to take advantage of variable packing

File: contracts/HubPool.sol [hunk 5]
```diff
@@ -105,14 +102,29 @@ contract HubPool is HubPoolInterface, Testable, Lockable, MultiCaller, Ownable {
...
     mapping(uint256 => CrossChainContract) public crossChainContracts;
 
+    mapping(address => uint256) public unclaimedAccumulatedProtocolFees;
+
+    // Whether the bundle proposal process is paused.
+    bool public paused;
+
     // WETH contract for Ethereum.
-    WETH9 public weth;
+    WETH9 public immutable weth;
```
Comment (nicholaspai) [inline contracts/HubPool.sol:111]: use `immutable` where possible

File: contracts/HubPool.sol [hunk 6]
```diff
@@ -603,60 +603,74 @@ contract HubPool is HubPoolInterface, Testable, Lockable, MultiCaller, Ownable {
 
     function executeRootBundle(
         uint256 chainId,
-        uint256[] memory bundleLpFees,
-        int256[] memory netSendAmounts,
-        int256[] memory runningBalances,
+        uint256[] calldata bundleLpFees,
```
Comment (nicholaspai) [inline contracts/HubPool.sol:606]: Use `calldata` for variables that will just be loaded into an in memory `PoolRebalance` struct

File: contracts/HubPool.sol [hunk 7]
```diff
@@ -603,60 +603,74 @@ contract HubPool is HubPoolInterface, Testable, Lockable, MultiCaller, Ownable {
...
-                }),
-                proof
-            ),
+            MerkleLib.verifyPoolRebalance(rootBundleProposal.poolRebalanceRoot, poolRebalanceLeaf, proof),
             "Bad Proof"
         );
 
         // Before interacting with a particular chain's adapter, ensure that the adapter is set.
-        require(address(crossChainContracts[chainId].adapter) != address(0), "No adapter for chain");
+        CrossChainContract memory _crossChainContracts = crossChainContracts[chainId];
```
Comment (nicholaspai) [inline contracts/HubPool.sol:633]: Reduce one SLOAD

File: contracts/HubPool.sol [hunk 8]
```diff
@@ -859,7 +873,8 @@ contract HubPool is HubPoolInterface, Testable, Lockable, MultiCaller, Ownable {
     ) internal {
         AdapterInterface adapter = crossChainContracts[chainId].adapter;
 
-        for (uint32 i = 0; i < l1Tokens.length; i++) {
+        uint32 length = uint32(l1Tokens.length);
```
Comment (nicholaspai) [inline contracts/HubPool.sol:876]: cache loop length

**Review by mrice32 — APPROVED** [review]
LGTM!

**Pull request merged**